### ISSUE TITLE
Save both checkpoints and log in R

### ISSUE THREAD

**Issue by Mikeriess** [dmlc/mxnet#5674]
Hi,

Im trying to store both a checkpoint for each iteration, as well as the log of the performance on the validation set. I've built a simple R-function, but it doesnt seem to work. Any hints on how to accomplish this?

> EpochEnd <- function(period_log,logger_log, prefix_model,period_model){
> mx.callback.save.checkpoint(prefix_model,period_model)
> mx.callback.log.train.metric(period_log, logger_log)
> }

> library(mxnet)
> data(BostonHousing, package="mlbench")
> train.ind = seq(1, 506, 3)
> train.x = data.matrix(BostonHousing[train.ind, -14])
> train.y = BostonHousing[train.ind, 14]
> test.x = data.matrix(BostonHousing[-train.ind, -14])
> test.y = BostonHousing[-train.ind, 14]
> 
> data <- mx.symbol.Variable("data")
> fc1 <- mx.symbol.FullyConnected(data, num_hidden=1)
> lro <- mx.symbol.LinearRegressionOutput(fc1)
> 
> logger <- mx.metric.logger$new()
> 
> model <- mx.model.FeedForward.create(
>   lro, 
>   X=train.x,
>   y=train.y,
>   eval.data=list(data=test.x, label=test.y),
>   array.layout="rowmajor",
>   ctx=mx.cpu(), 
>   num.round=5, 
>   array.batch.size=20,
>   learning.rate=2e-6, 
>   momentum=0.9, 
>   eval.metric=mx.metric.rmse,
>   epoch.end.callback = EpochEnd(1,logger,"boston",1)
>   )
> 

Somehow the log is saved, but not the checkpoints.

**Comment by Mikeriess** [comment]
Tried setting batch size = full dataset so that 1 batch = 1 epoch, and called log from the batch callback like this:

> logger <- mx.metric.logger$new()
> 
> model <- mx.model.FeedForward.create(
>   lro, 
>   X=train.x,
>   y=train.y,
>   eval.data=list(data=test.x, label=test.y),
>   array.layout="rowmajor",
>   ctx=mx.cpu(), 
>   num.round=90, 
>   array.batch.size=337,
>   learning.rate=2e-6, 
>   momentum=0.9, 
>   eval.metric=mx.metric.mse,
>   epoch.end.callback = mx.callback.save.checkpoint("boston"),
>   batch.end.callback = mx.callback.log.train.metric(1, logger) 
>   )
> 
> #see the log:
> logger$eval
> 
> #load a checkpointed model:
> Bestmodel <- mx.model.load("boston",90)

Which seems to be working, and thereby solved my problem!

**Comment by thirdwing** [comment]
@Mikeriess You can write a callback function like below:

```r
mx.callback.cus <- function(prefix, period, logger=NULL) {
  function(iteration, nbatch, env, verbose) {
    if (nbatch %% period == 0 && !is.null(env$metric)) {
      result <- env$metric$get(env$train.metric)
      if (nbatch != 0)
        if(verbose) cat(paste0("Batch [", nbatch, "] Train-", result$name, "=", result$value, "\n"))
      if (!is.null(logger)) {
        if (class(logger) != "mx.metric.logger") {
          stop("Invalid mx.metric.logger.")
        }
        logger$train <- c(logger$train, result$value)
        if (!is.null(env$eval.metric)) {
          result <- env$metric$get(env$eval.metric)
          if (nbatch != 0)
            cat(paste0("Batch [", nbatch, "] Validation-", result$name, "=", result$value, "\n"))
          logger$eval <- c(logger$eval, result$value)
        }
      }
    }
    if (iteration %% period == 0) {
      mx.model.save(env$model, prefix, iteration)
      if(verbose) cat(sprintf("Model checkpoint saved to %s-%04d.params\n", prefix, iteration))
    }
    return(TRUE)
  }
}


library(mxnet)
data(BostonHousing, package="mlbench")
train.ind = seq(1, 506, 3)
train.x = data.matrix(BostonHousing[train.ind, -14])
train.y = BostonHousing[train.ind, 14]
test.x = data.matrix(BostonHousing[-train.ind, -14])
test.y = BostonHousing[-train.ind, 14]
data <- mx.symbol.Variable("data")
fc1 <- mx.symbol.FullyConnected(data, num_hidden=1)
lro <- mx.symbol.LinearRegressionOutput(fc1)
mx.set.seed(0)

logger <- mx.metric.logger$new()
model <- mx.model.FeedForward.create(
  lro, X=train.x, y=train.y,
  eval.data=list(data=test.x, label=test.y),
  ctx=mx.cpu(), num.round=10, array.batch.size=20,
  learning.rate=2e-6, momentum=0.9, eval.metric=mx.metric.rmse,
  epoch.end.callback = mx.callback.cus("boston", 5, logger))
```

```
Start training with 1 devices
[1] Train-rmse=16.063282524034
[1] Validation-rmse=10.1766446093622
Model checkpoint saved to boston-0001.params
......
```

```
> head(logger$train)
[1] 16.063283 12.279238 11.198463 10.264524  9.497110  9.077337
> head(logger$eval)
[1] 10.176645 12.433178 10.330304  8.427604  8.445578  8.332255
```
